Add jemalloc support

ddn0 opened the pull request.

roshandathathri commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -2844,10 +2825,12 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
       reduce<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+      broadcast<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+    } else {
+      broadcast<writeSource, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
```
This is the same for if and else, so move it out.

roshandathathri commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -2863,20 +2846,13 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
           loopName);
-    } else {
+    }
+    if (!is_incoming_cut_) {
```
Move this if block before the previous if block. Always, reduce first and then broadcast.

roshandathathri commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -2912,10 +2888,15 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
+      broadcast<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
+          loopName);
+    } else {
+      broadcast<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
```
Common for both if and else block, so move it out.

roshandathathri commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -2912,10 +2888,15 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
-    if (!transposed || isVertexCut) {
-      reduce<writeDestination, readAny, SyncFnTy, BitsetFnTy, async>(loopName);
+    if (!is_incoming_cut_) {
+      reduce<writeDestination, readDestination, SyncFnTy, BitsetFnTy, async>(
```
Seems to be a typo. It's readAny for all functions within dst_to_any.

roshandathathri commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -2997,32 +2978,60 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
-          sync_dst_to_dst<SyncFnTy, BitsetFnTy, async>(loopName);
-        } else { // readAny
-          sync_dst_to_any<SyncFnTy, BitsetFnTy, async>(loopName);
+      if (transposed_) {
```
Add a comment saying "swap source and destination for read/write location (does not apply to any)"

roshandathathri reviewed (approved):
Yay! Thanks for doing this.

Except sync_dst_to_src(), rest LGTM. The other comments are minor nits.

gurbinder533 commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -3269,6 +3269,91 @@ class Partitioner {
+          [&](uint32_t n) {
+            out_indices_tmp[n] = dist_.large_array_out_indices_[n - 1];
+          },
+          galois::no_stats());
```
It is not exactly swapping actually.

roshandathathri replied:
I got that but the downstream code can be made to work with swapping if I understand right?

ddn0 committed:
build: Use KATANA_USE over KATANA_ENABLE

There might be some subtle nuance that is conveyed by "use" versus
"enable", but the distinction seems too fine to reliably maintain.

There doesn't seem to be a clear consensus among other projects either.
Here is a selection of options from related projects:

  CMAKE_USE_RELATIVE_PATHS
  CMAKE_ENABLE_TESTING
  LLVM_ENABLE_RTTI
  ARROW_USE_ASAN
  ARROW_WITH_ZLIB

Instead, just pick one form consistently over another. "Use" was chosen
because it was the form used by more recently added options to the
build.

Along the way, remove the "Katana" prefix from CMake files as it is
somewhat redundant with CMAKE_MODULE_PATH as paths in CMAKE_MODULE_PATH
will be searched first and then followed by default paths.

ddn0 committed:
libsupport: Add jemalloc support

ddn0 committed:
docs: Add ProgressTracer

patrickkenney9801 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -204,6 +209,22 @@ katana::ProgressSpan::LogError(
+    return;
+  }
+  static int stage;
+  SetTags({{"profile_stage", stage}});
```
I would make this a log since it would be better if developers don't write to the same tag on a span multiple times

patrickkenney9801 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -10,55 +10,64 @@
-/// However, logging and tagging existing spans are thread-safe in the JSONTracer
+/// Tracers do not currently support thread-local tracers or concurrency controls.
+/// Starting/Finishing span functions should only be used in a single-threaded context.
+/// However, logging and tagging existing spans is thread-safe.
```
Nit: this is plural since: logging and tagging are thread-safe

arthurp commented on libgalois/src/PageAlloc.cpp:
```diff
@@ -75,6 +64,32 @@ katana::allocSize() {
+  if (num == 0) {
+    return nullptr;
+  }
+  return malloc(num * hugePageSize);
```
Does using malloc here mean we are not using huge prayers in this mode? Or does jemalloc use them automatically? If the former is true, then we probably should print a warning and document that jamalloc is not as fast as our own allocator in some cases.

ddn0 replied:
jemalloc doesn't use huge pages explicitly though you can configure transparent huge pages; though there is some debate as to whether that is actually better for jemalloc: https://discuss.aerospike.com/t/disabling-transparent-huge-pages-thp-for-aerospike/5233

As for the warning,
1. I am not actually sure if with or without jemalloc is faster or slower so I don't think we have anything valuable to say in the documentation besides this may affect your performance in some way but isn't that kind of the point of changing `malloc`?
2. FWIW, the existing "you are not using huge pages" warning only appears in debug mode: https://github.com/ddn0/katana/blob/feature/jemalloc/libgalois/src/PageAlloc.cpp#L101

arthurp commented on docs/contributing/performance.rst:
```diff
@@ -112,3 +112,63 @@ Here is a rule of thumb to choose among these different options:
+
+.. warning::
+
+   The standard conda package for jemalloc does not enable profiling. You need
+   to use a conan build instead.
```
As we move away from conan and toward using conda more, we will need to fix this. https://katanagraph.atlassian.net/browse/KAT-1387

arthurp reviewed (approved):
One notable documentation/performance comment, but otherwise LGTM.

roshandathathri commented on docs/contributing/performance.rst:
```diff
@@ -112,3 +112,63 @@ Here is a rule of thumb to choose among these different options:
+
+``perf`` is a commonly available tool for profiling execution on Linux. You can
+collect samples from any execution with ``perf record <command line to
+profile>`` and then use ``perf report`` to view the report.
```
Not for this PR but it would be good to start and stop options in the code for perf.

roshandathathri reviewed (approved):
LGTM. Thanks!
